### ISSUE TITLE
BC/BSC support for Blockchain Balances

### ISSUE THREAD

**Issue by mariodigital** [rotki/rotki#1573]
Please add support for Binance Chain / Binance Smart Chain Blockchain addresses to automatically accumulate your balances under the dashboard.

**Comment by petermllrr** [comment]
🙏

**Comment by qqq-jump** [comment]
+1

**Comment by mariodigital** [comment]
Any day now 😅

**Comment by LefterisJP** [comment]
Hey guys. Thanks for +1 on this issue. We don't really know when developer resources can be dedicated to do this issue right.

Rotki is opensource and PRs are greatly appreciated.

The only open API I could find is etherscan's bscscan: https://bscscan.com/apis#accounts
But that's only for BNB balances.

**Comment by gufertum** [comment]
> But that's only for BNB balances.

AFAIK it's a clone of etherscan.io and therefore the apis look nearly the same. You can check BEP-20 Tokens just like ERC-20 on etherscan...

just my 2 cents

**Comment by LefterisJP** [comment]
> You can check BEP-20 Tokens just like ERC-20 on etherscan.

Yes but you need to know the token addresses to query them. We don't use etherscan's api to query ethereum token addresses in rotki. That would not work as there is over 10k tokens and the etherscan api accepts 1 token per request.

I guess would need a curated list of BSC token addresses etc, and probably figure out a nice way to do batch queries to not get rate limited by that api.

Or find a better api.

But as with any chain a proper solution would be to be able to connect to your own BSC node if you got one. But ofc step by step. This issue should just be for the first iteration.

I will repeat my call to action from above. We need help here. If you want this done soon consider opening a PR or finding someone who can do that is indeed the fastest way to see this in rotki.

If not we still want to see it in the app ofc. But we are a team of 2-3 people with very limited resources and a huge issue backlog. So no guarantees as to when this will happen.

**Comment by a7e8** [comment]
A good first step would be to just support BNB on the Binance Chain (not Smart Chain) using the explorer or by querying a node like the `bnbcli` does.

**Comment by MarcBerneman** [comment]
> Hey guys. Thanks for +1 on this issue. We don't really know when developer resources can be dedicated to do this issue right.
> 
> Rotki is opensource and PRs are greatly appreciated.
> 
> The only open API I could find is etherscan's bscscan: https://bscscan.com/apis#accounts
> But that's only for BNB balances.

Actually, you can use that API to inquire non-BNB tokens. The function you need is "Get BEP20-Token Account Balance for TokenContractAddress" (see [https://bscscan.com/apis#tokens](https://bscscan.com/apis#tokens)).  There is also a nice python module called bscsan-python that can interface with this API (see [https://github.com/pcko1/bscscan-python](https://github.com/pcko1/bscscan-python)).
I don't have time right now, but I might be willing to give this issue a shot once I do :)

**Comment by LefterisJP** [comment]
Separate issue for Binance Chain: https://github.com/rotki/rotki/issues/1573

**Comment by jkfliu** [comment]
I think the easiest way would be via this bscscan API:
**Get Address BEP20 Token Holding** (see https://bscscan.com/apis#APIpro)
The only problem is it's a 'Pro' API, meaning you'll need to pay (currently min $199/mth) to access it.

@LefterisJP if you've seen instructions on how to run a BSC light node, please do share (I've only been able to find instructions for Binance Chain light node - documentation is not great but I managed to get it up and running).

**Comment by LefterisJP** [comment]
> The only problem is it's a 'Pro' API, meaning you'll need to pay (currently min $199/mth) to access it.

Hehe that would make it really hard to actually work. As the user would need to buy the key for themselves. Which is an insane expense for a single user who just wants to see BSC. For reference our current price for the rotki premium subscription is $12.7/mth.

> if you've seen instructions on how to run a BSC light node,

I can't say I have. Another option may be something like covalent, or the graph.

**Comment by jkfliu** [comment]
Ok, I managed to get BSC balances using the Covalent API
GET /v1/{chain_id}/address/{address}/balances_v2/

Happy to try and help further but need some guidance please!

**Comment by LefterisJP** [comment]
Hey @jkfliu what would probaly help is to take a look at the work done here: https://github.com/rotki/rotki/pull/3242

That PR added support for Covalent API, and also support for the ability to add avalance addresses and query them along with all other rotki queries.

Needs quite a bit of refactoring around the common points of EVM based blockchain managers as per the comments on the review there but the idea is that.

**Comment by Spenhouet** [comment]
The Bscscan API actually also has a free mode with 5 queries per second: https://bscscan.com/apis
But another Option might be to directly use Web3 to query everything? In private I'm working on a small abstraction layer and until now I only needed Bscscan to get the ABI. The rest I could get via Web3.

It would be really great if the support can be added.

**Comment by LefterisJP** [comment]
Hey @Spenhouet thank you for chimin in!

We are working on a more generic EVM abstraction over assets and chains so we can handle all such chains together in a proper manner. Once that is done then we can easily add support for BSC and other such chains.

Until then you are also more than welcome to submit a PR to do an interim support for BSC using covalent as explained above: https://github.com/rotki/rotki/issues/1573#issuecomment-889123313

We need as much help from external contributors as possible :sweat_smile:

**Comment by Spenhouet** [comment]
> We are working on a more generic EVM abstraction over assets and chains

Could you share the issue this is tracked with?

**Comment by LefterisJP** [comment]
> > We are working on a more generic EVM abstraction over assets and chains
> 
> Could you share the issue this is tracked with?

Sure. It's this issue: https://github.com/rotki/rotki/issues/2701

An open draft PR and a long running branch on it are here:
https://github.com/rotki/rotki/pull/3743
https://github.com/rotki/rotki/tree/multichain_globaldb_hell

For easier chat and dev related questions btw we have a discord chat: https://discord.com/invite/aGCxHG7
Hop over there if you got more questions!

**Comment by thedelph** [comment]
Looking forward to this!

**Comment by immortaliseORdie** [comment]
so I guess for now we need to add the transactions manually?

**Comment by sarrek** [comment]
No update for BSC support ?

**Comment by yabirgb** [comment]
@sarrek We are in the process of adding optimism as we had a grant for that an other EVM chains will come after that since the process should be easier

**Comment by kaiqiangh** [comment]
Any update for BSC? Thanks

**Comment by MacDefender** [comment]
Any updates? Would love to use Rotki (and pay for Premium) but without BSC and Cronos Chain Support Rotki is not of much use at the moment.

**Comment by markxma1** [comment]
Is there some way to add BSC wallet data?
